Code/Engine/RendererCore/ShaderCompiler/Implementation/ShaderParser.cpp

```diff
@@ -1,163 +1,386 @@
+  {
+    ezUInt32 uiValueToken = uiCurToken;
 
-    const char* szNameStart = s.GetData();
-    while (s.IsValid() && IsIdentifier(s.GetCharacter()))
+    if (Accept(Tokens, uiCurToken, ezTokenType::String1, &uiValueToken) || Accept(Tokens, uiCurToken, ezTokenType::String2, &uiValueToken))
     {
-      ++s;
+      ezStringBuilder sValue = Tokens[uiValueToken]->m_DataView;
+      sValue.Trim("\"'");
+
+      return ezVariant(sValue.GetData());
     }
 
-    ezShaderParser::AttributeDefinition attributeDef;
-    attributeDef.m_sName = ezStringView(szNameStart, s.GetData());
+    if (Accept(Tokens, uiCurToken, ezTokenType::Integer, &uiValueToken))
+    {
+      ezString sValue = Tokens[uiValueToken]->m_DataView;
+
+      ezInt64 iValue = 0;
+      if (sValue.StartsWith_NoCase("0x"))
+      {
+        iValue = ezConversionUtils::ConvertHexStringToUInt32(sValue);
+      }
+      else
+      {
+        ezConversionUtils::StringToInt64(sValue, iValue);
+      }
 
-    SkipWhitespace(s);
+      return ezVariant(iValue);
+    }
 
-    if (!s.IsValid() || s.GetCharacter() != '(')
-      return;
+    if (Accept(Tokens, uiCurToken, ezTokenType::Float, &uiValueToken))
+    {
+      ezString sValue = Tokens[uiValueToken]->m_DataView;
+
+      double fValue = 0;
+      ezConversionUtils::StringToFloat(sValue, fValue);
 
-    ++s; // skip (
-    int braces = 0;
+      return ezVariant(fValue);
+    }
 
-    const char* szValueStart = s.GetData();
-    while (s.IsValid())
+    if (Accept(Tokens, uiCurToken, "true", &uiValueToken) || Accept(Tokens, uiCurToken, "false", &uiValueToken))
     {
-      if (s.GetCharacter() == '(')
+      bool bValue = Tokens[uiValueToken]->m_DataView == "true";
+      return ezVariant(bValue);
+    }
+
+    auto& dataView = Tokens[uiCurToken]->m_DataView;
+    if (Tokens[uiCurToken]->m_iType == ezTokenType::Identifier &&
+        ezStringUtils::IsValidIdentifierName(dataView.GetStartPosition(), dataView.GetEndPosition()))
+    {
+      // complex type constructor
+      const ezRTTI* pType = nullptr;
+      if (!s_NameToTypeTable.TryGetValue(dataView, pType))
+      {
+        ezLog::Error("Invalid type name '{}'", dataView);
+        return ezVariant();
+      }
+
+      ++uiCurToken;
+      Accept(Tokens, uiCurToken, "(");
+
+      ezHybridArray<ezVariant, 8> constructorArgs;
+
+      while (!Accept(Tokens, uiCurToken, ")"))
       {
-        braces++;
+        ezVariant value = ParseValue(Tokens, uiCurToken);
+        if (value.IsValid())
+        {
+          constructorArgs.PushBack(value);
+        }
+        else
+        {
+          ezLog::Error("Invalid arguments for constructor '{}'", pType->GetTypeName());
+          return EZ_FAILURE;
+        }
 
-        ++s; // skip (
-        szValueStart = s.GetData();
+        Accept(Tokens, uiCurToken, ",");
       }
-      else
+
+      // find matching constructor
+      auto& functions = pType->GetFunctions();
+      for (auto pFunc : functions)
```

Comment:
@Sanakan8472 Is this the right way to use constructor reflection? It seems to work ;)